From b2738ffb2fdd2550ececb74f76f75583c43c8b59 Mon Sep 17 00:00:00 2001
From: "Ignacio R. Morelle" <shadowm@wesnoth.org>
Date: Tue, 16 Jun 2015 23:08:52 -0300
Subject: [PATCH] Use looks_like_pbl() to disallow .pbl file inclusion (bug
 #23504)

This function is implemented using case-insensitive pattern matching,
unlike filesystem::ends_with(). I missed this when writing my original
fix, so the vulnerability still applied to .pbl files on a
case-insensitive filesystem (e.g. NTFS and FAT* on Windows) by using
different case to bypass the check.
---
 changelog                | 3 ++-
 players_changelog        | 3 ++-
 src/filesystem.cpp       | 2 +-
 src/filesystem_boost.cpp | 2 +-
 4 files changed, 6 insertions(+), 4 deletions(-)

diff --git a/changelog b/changelog
index 7f1d62c8f1fa..9554b610b1e8 100644
--- a/changelog
+++ b/changelog
@@ -1,6 +1,7 @@
 Version 1.13.0+dev:
  * Security fixes:
-   * Disallowed inclusion of .pbl files from WML (bug #23504).
+   * Disallowed inclusion of .pbl files from WML (bug #23504), independent of
+     file name case.
  * AI:
    * Fast Micro AI: exclude hidden enemies from attack evaluation by default;
      add optional key attack_hidden_enemies to override this
diff --git a/players_changelog b/players_changelog
index 3ab0758dc62e..0a1019e2e0e1 100644
--- a/players_changelog
+++ b/players_changelog
@@ -4,7 +4,8 @@ changelog: https://github.com/wesnoth/wesnoth/blob/master/changelog
 
 Version 1.13.0+dev:
  * Security fixes:
-   * Disallowed inclusion of .pbl files from WML (bug #23504).
+   * Disallowed inclusion of .pbl files from WML (bug #23504), independent of
+     file name case.
 
  * Campaigns:
    * Dead Water:
diff --git a/src/filesystem.cpp b/src/filesystem.cpp
index 7d5adb60d78c..01cd577b1fb8 100644
--- a/src/filesystem.cpp
+++ b/src/filesystem.cpp
@@ -1008,7 +1008,7 @@ std::string get_wml_location(const std::string &filename, const std::string &cur
 		return result;
 	}
 
-	if (ends_with(filename, ".pbl")) {
+	if (looks_like_pbl(filename)) {
 		ERR_FS << "Illegal path '" << filename << "' (.pbl files are not allowed)." << std::endl;
 		return result;
 	}
diff --git a/src/filesystem_boost.cpp b/src/filesystem_boost.cpp
index 21b582159e11..08cbd2666f6e 100644
--- a/src/filesystem_boost.cpp
+++ b/src/filesystem_boost.cpp
@@ -1000,7 +1000,7 @@ static bool is_legal_file(const std::string &filename)
 		return false;
 	}
 
-	if (ends_with(filename, ".pbl")) {
+	if (looks_like_pbl(filename)) {
 		ERR_FS << "Illegal path '" << filename << "' (.pbl files are not allowed)." << std::endl;
 		return false;
 	}
